.travis.yml

```diff
@@ -20,13 +20,19 @@ jobs:
     - stage: test
       jdk: openjdk11
       script: mvn verify -P-spotless-apply --toolchains .travis-toolchains.xml
-      env: VERIFY=true
+      env:
+      - VERIFY=true
+      # Remove this when the messages.cucumber.io domain is available
+      - CUCUMBER_MESSAGE_STORE_URL=https://b1f2yk58d8.execute-api.eu-west-3.amazonaws.com
```

Comment:
The presence of this environment variable is not what triggers the publishing. This just overrides the hard-coded `https://messages.cucumber.io` because that subdomain has not yet been set up.

Publishing is triggered by setting `publish=true` in the JUnit annotation, or by passing `--publish` on the CLI.

We have changed Cucumber so that it can publish results/reports to an external service. We're going to encourage users to do this. If we're encouraging users to do this I see no problem with doing it ourselves, eating our own dog food.

Can you elaborate on why you don't think we should publish results?

Reply:
There are are too many downsides to publishing reports as part of the build. I'd roughly group them into technical and legal issues.

1. If I were to run the build locally I now must have an internet connection. I've developed much of Cucumber in public transport without internet or on a metered connection. So this is generally undesirable.

2. The test itself doesn't assert anything but the fact that uploading resulted in a 200 status code. We've already got this more or less covered by the unit tests for the `UrlOutputStream` so the only unique thing that tests add is testing the connection the reports API which isn't part of Cucumber.

3. A dependency on an external service is a possibility for failure. Since this service isn't strictly necessary to build and or publish Cucumber JVM it should be eliminated from the build. This removes a possible cause of flakes.

4. If you do design an integration tests it might be worth doing so properly. This means using multiple versions of Cucumber JVM and using the CCK rather then a fairly minimal example.

Then from a legal and risk perspective:

1. It is unlikely that random contributors will have agreed to the T&Cs of the reports API.

2. Software should be safe by default. We should prevent a scenario where someone by copying an example unwillingly and/or unwittingly uploads proprietary information to a third party service. The liabilities and contractual penalties involved can be quite significant.

Reply:
Excellent points. @vincent-psarga and I had a chat about this today, and propose:

The `PublishFormatter` is enabled by either specifying `--publish` OR by defining `CUCUMBER_REPORTS_TOKEN`
We'll make it clear to users that `--publish` is only meant for "testing the waters" of Cucumber Reports, and should not be committed to git. Reports published this way will be open to anyone who knows the URL. Reports will be automatically deleted after a certain delay (1h, 6h, 24h, not sure) unless they are *claimed*.

When someone views a report published with `--publish` they will be able to *claim* that report by linking it to a GitHub repo. They'll authenticate with OAuth and then select the corresponding GitHub repo. After doing this they will see instructions to define an environment variable with a token in their CI environment: `CUCUMBER_REPORTS_TOKEN=xxxxxxx`. The presence of this environment variable will trigger the `PublishFormatter` and send the token as the `CUCUMBER_REPORTS_TOKEN` HTTP header. This will cause the report to be automatically claimed against the corresponding git repo.

Claimed reports can only be accessed by users who are authenticated via GitHub and have read permission to the corresponding repo.

Now onto your comments:

Technical:

1. This should be solved by the above. People hopefully won't commit their `--publish` / `publish = true` config, and if `CUCUMBER_REPORTS_TOKEN` is not defined, no publishing will happen

2. We'll later set `CUCUMBER_REPORTS_TOKEN` as an environment in CI (encrypted env var on Travis, or we can migrate to CircleCI or GitHub Actions where it can be set outside git.

3. The Cucumber Reports service is going to be a central part of the Cucumber ecosystem. It's our best shot at converting Cucumber Open users to paid products, and SmartBear's investment in Cucumber Open is contingent on this. We understand that some users may not want to publish reports, which is why we're trying to make it easy to opt out. It is however important that we gain confidence that reports publishing works as expected, which is why we want to verify this in the Cucumber-JVM build.

4. I disagree. Publishing reports from the CI environment seems unproblematic to me. Since the reports service is hosted by AWS we don't expect much downtime.

Legal:

1. The reports API is not collecting any [PII](https://en.wikipedia.org/wiki/Personal_data) from people contributing to Cucumber Open. The reports will only be published on the CI server using a `CUCUMBER_REPORTS_TOKEN` associated with the `cukebot@cucumber.io` GitHub user.

2. This is mitigated by deleting reports that are not claimed within a timeframe.